Source/JavaScriptCore/runtime/JSBigInt.cpp

```diff
@@ -473,6 +475,55 @@ class Int32BigIntImpl {
     int32_t m_value;
 };
 
+class Int64BigIntImpl {
+public:
+    explicit Int64BigIntImpl(int64_t value)
+        : m_value(value)
+        , m_sign(value < 0)
+    {
+    }
+
+    explicit Int64BigIntImpl(uint64_t value)
+        : m_value(value)
+        , m_sign(false)
+    { }
+
+    ALWAYS_INLINE bool isZero() { return !m_value; }
+    ALWAYS_INLINE bool sign() { return m_sign; }
+    ALWAYS_INLINE unsigned length()
+    {
+#if CPU(REGISTER64)
+        unsigned nonZeroLength = 1;
+#else
+        ASSERT(sizeof(JSBigInt::Digit) == 32);
+        unsigned nonZeroLength = 2;
+#endif
+        return isZero() ? 0 : nonZeroLength;
+    }
+    ALWAYS_INLINE JSBigInt::Digit digit(unsigned i)
+    {
+        ASSERT(length());
+#if CPU(REGISTER64)
+        ASSERT_UNUSED(i, !i);
+        if (sign())
+            return static_cast<JSBigInt::Digit>(~(m_value - 1));
+        return m_value;
+#else
+        ASSERT(!i || i == 1);
+        ASSERT(sizeof(JSBigInt::Digit) == 32);
+        uint64_t absoluteBits = m_value;
+        if (sign())
+            absoluteBits = ~(m_value - 1);
+        return static_cast<JSBigInt::Digit>(absoluteBits >> (32 * i));
+#endif
+    }
+
+private:
+    friend struct JSBigInt::ImplResult;
+    uint64_t m_value;
+    bool m_sign;
+};
```

Comment:
I'd do 
```c++
// In WTF/MathExtras.h:

// For use in places where we could negate std::numeric_limits<T>::min and would like to avoid UB.
template<std::integral T>
ALWAYS_INLINE T negate(T v)
{
    return ~static_cast<std::make_unsigned_t<T>>(v) + 1;
}

// Here:

class Int64BigIntImpl {
public:
    explicit Int64BigIntImpl(int64_t value)
        : m_bits(value)
        , m_sign(value < 0)
    {
    }

    explicit Int64BigIntImpl(uint64_t value)
        : m_bits(value)
        , m_sign(false)
    { }

    static constexpr unsigned numDigits = isRegister64Bit() ? 1 : 2;

    ALWAYS_INLINE bool isZero() { return !m_value; }
    ALWAYS_INLINE bool sign() { return m_sign; }
    ALWAYS_INLINE unsigned length()
    {
        return isZero() ? 0 : numDigits;
    }
    ALWAYS_INLINE JSBigInt::Digit digit(unsigned i)
    {
        ASSERT_UNUSED(i, i < length());
#if CPU(REGISTER64)
        if (sign())
            return static_cast<JSBigInt::Digit>(negate(m_value));
        return m_value;
#else
        if (sign()) {
            return static_cast<JSBigInt::Digit>(negate(m_value) >> (32 * i));
        }
        return static_cast<JSBigInt::Digit>(m_value >> (32 * i));
#endif
    }

private:
    friend struct JSBigInt::ImplResult;
    uint64_t m_value;
    bool m_sign;
};
```